Added possibility to define enums as objects to carry a value/label pair

emiloberg opened the pull request:
This PR enables you to define enums as objects (as well as todays arrays). This way you can supply a value/label pair. Example:

```
street_type: {
  type: 'string',
  enum: {
    st: 'Street',
    ave: 'Avenue',
    blvd: 'Boulevard'
  }
}
```

Which is equivalent to defining the enum like this in tcomb:

```
t.enums({
    st: 'Street',
    ave: 'Avenue',
    blvd: 'Boulevard'
}, 'Street');
```

No breaking changes made. Valueless arrays can still be used:

```
street_type: {
  type: 'string',
  enum: [ 'Street', 'Avenue', 'Boulevard']
}
```

Added custom `isArray` method to keep compatible with <=IE8

gcanti commented:
Hi @emiloberg,

Thanks for your PR, I'll gladly merge it. Just 2 observations:

> Added custom isArray method to keep compatible with <=IE8

What about using `t.Array.is` instead?

Also, could you please add a short note in the README.md file documenting your useful addition?

emiloberg commented:
_Of course_ I should've used `t.Array.is`! Changed and squashed. 

About documentation. I didn't add it to the README as it doesn't have any other documentation about the JSON structure. I was looking for a Github repo for the [JSON Schema to tcomb playground](https://gcanti.github.io/resources/json-schema-to-tcomb/playground/playground.html) but I guess there isn't one?

I was thinking  that changing example _3. Enums_ to the code below would be a good way to document it.

```
{
  "type": "object",
  "properties": {
    "number":      { "type": "number" },
    "street_name": { "type": "string" },
    "street_type": { 
      "type": "string",
      "enum": ["Street", "Avenue", "Boulevard"]
    },
    "country": { 
      "type": "string",
      "enum": {
        us: 'United States',
        it: 'Italy',
        se: 'Sweden'
      }
    }
  }
}
```

That said, I'll happily add a few lines to the README if you want me to. Just say and I'll fix!

gcanti commented:
> I didn't add it to the README as it doesn't have any other documentation about the JSON structure

Yep. Roughly the documention assumes that the accepted JSON schema is standard. I'd document just the additions (example the `registerType` API).

> I'll happily add a few lines to the README if you want me to

Yes please. Nothing fancy, just a note (and maybe a simple example) for future users, thanks!

emiloberg commented:
Done!

gcanti commented:
Released in https://github.com/gcanti/tcomb-json-schema/releases/tag/v0.2.4
